### DISCUSSION TITLE
Access Extended Page Data in Enhance Client App

### DISCUSSION THREAD

**Discussion by drewcook** [vuepress/core#487] [Q&A] [answered]
Title sums it up.  I am extending the page data via a plugin.  Let's say I add a new field `meta` to the Page object.  I can access that just fine at runtime at the Vue layer, say in `Layout.vue` without issue.  However, I need to also access this meta data I've created when I enhance the client app.  Since there's no feature built into VuePress that allows us to customize the `document.title` format, I'm going this route as a solution.

I want to use the metadata that I've put into `meta` via `extendsPageData` in a plugin to achieve this, however, on initial load, this is undefined.  The second time it is called, it _is_ defined.  What is the sequence of events that is happening to where it is undefined at first but then defined.  At what point is my plugin's `extendsPageData` completing and at what point is `defineClientAppEnhance` being called?  Is there a race condition occurring?

```ts
// plugin.ts
const myPlugin = {
  name: 'myPlugin',
  extendsPageData: (page) => {
  const meta = 'foobar'
    return { meta }
  },
};

// Layout.vue
import { usePageData } from '@vuepress/client'

const { value: { meta } } = usePageData()
console.log(meta) // 'foobar'

// enhanceClientApp.ts
import { usePageData } from '@vuepress/client'

// ideally, I'd like to do something like the following, since there's no easy way to do this via any API in VuePress
function generateDynamicTitle(meta) {
  if (document) document.title = `${meta} - My Site Name | ${otherMeta.foo}`
}

export default defineClientAppEnhance(({ app, router, siteData }) => {
  const { value: { meta } } = usePageData()
  console.log(meta) // undefined
  generateDynamicTitle(meta)

  router.afterEach(() => {
    console.log(meta) // 'foobar'
    generateDynamicTitle(meta)
  }
}
```

**Comment by laoergege** [comment]
Oh, I also meet. Because the data is async load. In my mind, the pageData should be tracked, which is only read.

**Comment by laoergege** [comment]
I commit an issue, but nobody respond.

https://github.com/vuepress/vuepress-next/issues/477#issue-1028301601

**Comment by meteorlxy** [comment, marked as answer]
There is a previous discussion related to that: https://github.com/vuepress/vuepress-next/discussions/72

**Reply by laoergege** [reply]
Thank you. I finally got my reason:hand_over_mouth:

**Reply by laoergege** [reply]
I use @vuepress/client in another project in which I write a vuepress plugin. The webpack finally bundled two different @vuepress/client codes. The correct way is to use the $page on the app instance.

**Reply by laoergege** [reply]
But I still think usePageData should return a computed value so that it can be tracked.
